codegen/smithy-go-codegen/src/main/java/software/amazon/smithy/go/codegen/protocol/aws/DeserializeMiddleware.java

```diff
@@ -0,0 +1,219 @@
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package software.amazon.smithy.go.codegen.protocol.aws;
+
+import static software.amazon.smithy.go.codegen.GoWriter.goTemplate;
+import static software.amazon.smithy.go.codegen.SmithyGoDependency.SMITHY_HTTP_TRANSPORT;
+import static software.amazon.smithy.go.codegen.protocol.ProtocolUtil.hasEventStream;
+import static software.amazon.smithy.go.codegen.server.protocol.JsonDeserializerGenerator.getDeserializerName;
+
+import software.amazon.smithy.go.codegen.GoStdlibTypes;
+import software.amazon.smithy.go.codegen.GoWriter;
+import software.amazon.smithy.go.codegen.SmithyGoDependency;
+import software.amazon.smithy.go.codegen.integration.ProtocolGenerator;
+import software.amazon.smithy.model.shapes.OperationShape;
+import software.amazon.smithy.model.shapes.StructureShape;
+import software.amazon.smithy.utils.MapUtils;
+
+public class DeserializeMiddleware {
+
+    public static final String SMITHY_PROTOCOL_NAME = "awsJson10";
+    protected final ProtocolGenerator.GenerationContext ctx;
+    protected final OperationShape operation;
+    protected final GoWriter writer;
+
+    protected final StructureShape input;
+    protected final StructureShape output;
+
+    private String deserialName;
+
+    public DeserializeMiddleware(
+            ProtocolGenerator.GenerationContext ctx, OperationShape operation, GoWriter writer
+    ) {
+        this.ctx = ctx;
+        this.operation = operation;
+        this.writer = writer;
+
+        this.input = ctx.getModel().expectShape(operation.getInputShape(), StructureShape.class);
+        this.output = ctx.getModel().expectShape(operation.getOutputShape(), StructureShape.class);
+    }
+
+    public static String getMiddlewareName(OperationShape operation) {
+        return "awsAwsjson10_deserializeOp" + operation.toShapeId().getName();
+    }
+
+    public GoWriter.Writable generate() {
+        deserialName = getMiddlewareName(operation);
+
+        return goTemplate("""
+
+            type $opName:L struct{
+            }
+
+            func (op *$opName:L) ID() string {
+                return "OperationDeserializer"
+            }
+
+            $handleSerialize:W
+
+            """,
+                MapUtils.of(
+                        "opName", deserialName,
+                        "handleSerialize", generateHandleDeserialize()
+                ));
+    }
+
+    private GoWriter.Writable generateHandleDeserialize() {
+        return goTemplate(
+                """
+
+                        func (op *$opName:L) HandleDeserialize (ctx $context:T, in $input:T, next $handler:T) (
```

Comment:
If you're tired of writing these out by hand, take a look at the APIs in `GoStackStepMiddlewareGenerator` - e.g. `createDeserializeStepMiddleware`